{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# List of pathogenic variants for PD\n",
    "Created: 2018-02-05    \n",
    "By: Hirotaka\n",
    "\n",
    "1. Reference clinvar files (2 versions)\n",
    "2. Pull pathogeic/likely_pathogenic/Conflicting_interpretations_of_pathogenicity\n",
    "3. Create the combined list of variants of interest from 2 versions\n",
    "4. Annotate these variants with 2 versions of clinvar\n",
    "5. Process the two annovar outputs into one."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "-----------------------------------------------------------------\n",
      "NOTICE: Processing operation=g protocol=refGene\n",
      "\n",
      "NOTICE: Running with system command <annotate_variation.pl -geneanno -buildver hg19 -dbtype refGene -outfile clinvar_20170501.refGene -exonsort _combined.avinput /fdb/annovar/2018-04-16/hg19>\n",
      "NOTICE: Output files were written to clinvar_20170501.refGene.variant_function, clinvar_20170501.refGene.exonic_variant_function\n",
      "NOTICE: Reading gene annotation from /fdb/annovar/2018-04-16/hg19/hg19_refGene.txt ... Done with 69265 transcripts (including 16521 without coding sequence annotation) for 28038 unique genes\n",
      "NOTICE: Processing next batch with 130 unique variants in 130 input lines\n",
      "NOTICE: Reading FASTA sequences from /fdb/annovar/2018-04-16/hg19/hg19_refGeneMrna.fa ... Done with 69 sequences\n",
      "WARNING: A total of 432 sequences will be ignored due to lack of correct ORF annotation\n",
      "-----------------------------------------------------------------\n",
      "NOTICE: Processing operation=r protocol=cytoBand\n",
      "\n",
      "NOTICE: Running with system command <annotate_variation.pl -regionanno -dbtype cytoBand -buildver hg19 -outfile clinvar_20170501 _combined.avinput /fdb/annovar/2018-04-16/hg19>\n",
      "NOTICE: Output file is written to clinvar_20170501.hg19_cytoBand\n",
      "NOTICE: Reading annotation database /fdb/annovar/2018-04-16/hg19/hg19_cytoBand.txt ... Done with 862 regions\n",
      "NOTICE: Finished region-based annotation on 130 genetic variants\n",
      "-----------------------------------------------------------------\n",
      "NOTICE: Processing operation=f protocol=avsnp150\n",
      "\n",
      "NOTICE: Running system command <annotate_variation.pl -filter -dbtype avsnp150 -buildver hg19 -outfile clinvar_20170501 _combined.avinput /fdb/annovar/2018-04-16/hg19>\n",
      "NOTICE: Variants matching filtering criteria are written to clinvar_20170501.hg19_avsnp150_dropped, other variants are written to clinvar_20170501.hg19_avsnp150_filtered\n",
      "NOTICE: Processing next batch with 130 unique variants in 130 input lines\n",
      "NOTICE: Database index loaded. Total number of bins is 28258790 and the number of bins to be scanned is 104\n",
      "NOTICE: Scanning filter database /fdb/annovar/2018-04-16/hg19/hg19_avsnp150.txt...Done\n",
      "-----------------------------------------------------------------\n",
      "NOTICE: Processing operation=f protocol=clinvar_20170501\n",
      "NOTICE: Finished reading 5 column headers for '-dbtype clinvar_20170501'\n",
      "\n",
      "NOTICE: Running system command <annotate_variation.pl -filter -dbtype clinvar_20170501 -buildver hg19 -outfile clinvar_20170501 _combined.avinput /fdb/annovar/2018-04-16/hg19 -otherinfo>\n",
      "NOTICE: the --dbtype clinvar_20170501 is assumed to be in generic ANNOVAR database format\n",
      "NOTICE: Variants matching filtering criteria are written to clinvar_20170501.hg19_clinvar_20170501_dropped, other variants are written to clinvar_20170501.hg19_clinvar_20170501_filtered\n",
      "NOTICE: Processing next batch with 130 unique variants in 130 input lines\n",
      "NOTICE: Database index loaded. Total number of bins is 38585 and the number of bins to be scanned is 84\n",
      "NOTICE: Scanning filter database /fdb/annovar/2018-04-16/hg19/hg19_clinvar_20170501.txt...Done\n",
      "-----------------------------------------------------------------\n",
      "NOTICE: Multianno output file is written to clinvar_20170501.hg19_multianno.txt\n",
      "-----------------------------------------------------------------\n",
      "NOTICE: Processing operation=g protocol=refGene\n",
      "\n",
      "NOTICE: Running with system command <annotate_variation.pl -geneanno -buildver hg19 -dbtype refGene -outfile clinvar_20180603.refGene -exonsort _combined.avinput /fdb/annovar/2018-04-16/hg19>\n",
      "NOTICE: Output files were written to clinvar_20180603.refGene.variant_function, clinvar_20180603.refGene.exonic_variant_function\n",
      "NOTICE: Reading gene annotation from /fdb/annovar/2018-04-16/hg19/hg19_refGene.txt ... Done with 69265 transcripts (including 16521 without coding sequence annotation) for 28038 unique genes\n",
      "NOTICE: Processing next batch with 130 unique variants in 130 input lines\n",
      "NOTICE: Reading FASTA sequences from /fdb/annovar/2018-04-16/hg19/hg19_refGeneMrna.fa ... Done with 69 sequences\n",
      "WARNING: A total of 432 sequences will be ignored due to lack of correct ORF annotation\n",
      "-----------------------------------------------------------------\n",
      "NOTICE: Processing operation=r protocol=cytoBand\n",
      "\n",
      "NOTICE: Running with system command <annotate_variation.pl -regionanno -dbtype cytoBand -buildver hg19 -outfile clinvar_20180603 _combined.avinput /fdb/annovar/2018-04-16/hg19>\n",
      "NOTICE: Output file is written to clinvar_20180603.hg19_cytoBand\n",
      "NOTICE: Reading annotation database /fdb/annovar/2018-04-16/hg19/hg19_cytoBand.txt ... Done with 862 regions\n",
      "NOTICE: Finished region-based annotation on 130 genetic variants\n",
      "-----------------------------------------------------------------\n",
      "NOTICE: Processing operation=f protocol=avsnp150\n",
      "\n",
      "NOTICE: Running system command <annotate_variation.pl -filter -dbtype avsnp150 -buildver hg19 -outfile clinvar_20180603 _combined.avinput /fdb/annovar/2018-04-16/hg19>\n",
      "NOTICE: Variants matching filtering criteria are written to clinvar_20180603.hg19_avsnp150_dropped, other variants are written to clinvar_20180603.hg19_avsnp150_filtered\n",
      "NOTICE: Processing next batch with 130 unique variants in 130 input lines\n",
      "NOTICE: Database index loaded. Total number of bins is 28258790 and the number of bins to be scanned is 104\n",
      "NOTICE: Scanning filter database /fdb/annovar/2018-04-16/hg19/hg19_avsnp150.txt...Done\n",
      "-----------------------------------------------------------------\n",
      "NOTICE: Processing operation=f protocol=clinvar_20180603\n",
      "NOTICE: Finished reading 5 column headers for '-dbtype clinvar_20180603'\n",
      "\n",
      "NOTICE: Running system command <annotate_variation.pl -filter -dbtype clinvar_20180603 -buildver hg19 -outfile clinvar_20180603 _combined.avinput /fdb/annovar/2018-04-16/hg19 -otherinfo>\n",
      "NOTICE: the --dbtype clinvar_20180603 is assumed to be in generic ANNOVAR database format\n",
      "NOTICE: Variants matching filtering criteria are written to clinvar_20180603.hg19_clinvar_20180603_dropped, other variants are written to clinvar_20180603.hg19_clinvar_20180603_filtered\n",
      "NOTICE: Processing next batch with 130 unique variants in 130 input lines\n",
      "NOTICE: Database index loaded. Total number of bins is 44107 and the number of bins to be scanned is 86\n",
      "NOTICE: Scanning filter database /fdb/annovar/2018-04-16/hg19/hg19_clinvar_20180603.txt...Done\n",
      "-----------------------------------------------------------------\n",
      "NOTICE: Multianno output file is written to clinvar_20180603.hg19_multianno.txt\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "OLD=clinvar_20170501\n",
    "NEW=clinvar_20180603\n",
    "OLDF=\"hg19_\"$OLD.txt\n",
    "NEWF=\"hg19_\"$NEW.txt\n",
    "\n",
    "grep athogenic $ANNOVAR_DATA/hg19/$OLDF | \\\n",
    " grep 'arkinson' | \\\n",
    " grep -v 'Wolff' | \\\n",
    " grep -v 'Infantile' \\\n",
    " > $OLD.avinput\n",
    "\n",
    "grep athogenic $ANNOVAR_DATA/hg19/$NEWF | \\\n",
    " grep 'arkinson' | \\\n",
    " grep -v 'Wolff' | \\\n",
    " grep -v 'Infantile' \\\n",
    " > $NEW.avinput\n",
    " \n",
    "cat $OLD.avinput $NEW.avinput | cut -f1-5 | sort | uniq > _combined.avinput\n",
    "\n",
    "table_annovar.pl _combined.avinput $ANNOVAR_DATA/hg19 \\\n",
    " -buildver hg19 \\\n",
    " -out $OLD \\\n",
    " -remove \\\n",
    " -protocol refGene,cytoBand,avsnp150,$OLD \\\n",
    " -operation g,r,f,f\n",
    " \n",
    "table_annovar.pl _combined.avinput $ANNOVAR_DATA/hg19 \\\n",
    " -buildver hg19 \\\n",
    " -out $NEW \\\n",
    " -remove \\\n",
    " -protocol refGene,cytoBand,avsnp150,$NEW \\\n",
    " -operation g,r,f,f\n",
    "\n",
    "table_annovar.pl _combined.avinput $ANNOVAR_DATA/hg19 \\\n",
    " -buildver hg19 \\\n",
    " -out freq \\\n",
    " -remove \\\n",
    " -protocol popfreq_max_20150413 \\\n",
    " -operation f"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "new = pd.read_table('clinvar_20180603.hg19_multianno.txt')\n",
    "def get_info(x):\n",
    "    if x['Func.refGene'] != 'exonic':\n",
    "        ans = x['Func.refGene']\n",
    "    else:\n",
    "        t = x['AAChange.refGene']\n",
    "        ans = [y for x,y in enumerate(t.split(':')) if ('p.' in y )and (',' not in y)][0]\n",
    "    return ans\n",
    "new['AAchange'] = new.apply(get_info, 1)\n",
    "old = pd.read_table('clinvar_20170501.hg19_multianno.txt')\n",
    "old['AAchange'] = old.apply(get_info, 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [],
   "source": [
    "old_derived = old[['avsnp150', 'Chr', 'Start', 'End', 'Ref', 'Alt', 'Gene.refGene', 'AAchange', 'CLINSIG', 'CLNDBN']]\n",
    "new_derived = new[['avsnp150', 'Chr', 'Start', 'End', 'Ref', 'Alt', 'Gene.refGene', 'AAchange', 'CLNSIG', 'CLNDN']]\n",
    "df = pd.merge(old_derived, new_derived, how='outer', \n",
    "              on=['avsnp150', 'Chr', 'Start', 'End', 'Ref', 'Alt', 'Gene.refGene', 'AAchange'],\n",
    "              suffixes=('_old', '_new'))\n",
    "def get_ChrOrder(x):\n",
    "    if x.Chr == 'X':\n",
    "        ans = 23\n",
    "    elif x.Chr == 'Y':\n",
    "        ans = 24\n",
    "    elif x.Chr == 'MT':\n",
    "        ans = 25\n",
    "    else:\n",
    "        ans = int(x.Chr)\n",
    "    return ans\n",
    "df['ChrOrder']=df.apply(get_ChrOrder, 1)\n",
    "df = df.sort_values(['ChrOrder', 'Start'])\n",
    "df = df.rename(columns={'CLINSIG':'CLNSIG_1705', 'CLNDBN':'CLNDBN_1705', 'CLNSIG':'CLNSIG_1806', 'CLNDN':'CLNDBN_1806'})\n",
    "df = df[['avsnp150', 'Chr', 'Start', 'End', 'Ref', 'Alt', 'Gene.refGene', 'AAchange', \n",
    "         'CLNSIG_1705', 'CLNSIG_1806', 'CLNDBN_1705', 'CLNDBN_1806']]\n",
    "df.to_csv('clinvar_1806_1705.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "rm *.avinput"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>avsnp150</th>\n",
       "      <th>Chr</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Ref</th>\n",
       "      <th>Alt</th>\n",
       "      <th>Gene.refGene</th>\n",
       "      <th>AAchange</th>\n",
       "      <th>CLNSIG_1705</th>\n",
       "      <th>CLNSIG_1806</th>\n",
       "      <th>CLNDBN_1705</th>\n",
       "      <th>CLNDBN_1806</th>\n",
       "      <th>SNP</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs76763715</td>\n",
       "      <td>1</td>\n",
       "      <td>155205634</td>\n",
       "      <td>155205634</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>GBA</td>\n",
       "      <td>p.N409S</td>\n",
       "      <td>Pathogenic|other|other|Pathogenic|Pathogenic|P...</td>\n",
       "      <td>Pathogenic/Likely_pathogenic,_risk_factor</td>\n",
       "      <td>Gaucher's_disease,_type_1|Parkinson_disease,_l...</td>\n",
       "      <td>Dementia,_Lewy_body,_susceptibility_to|Rigidit...</td>\n",
       "      <td>rs76763715</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs34637584</td>\n",
       "      <td>12</td>\n",
       "      <td>40734202</td>\n",
       "      <td>40734202</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>LRRK2</td>\n",
       "      <td>p.G2019S</td>\n",
       "      <td>Pathogenic|Pathogenic</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Parkinson_disease_8,_autosomal_dominant|not_pr...</td>\n",
       "      <td>Inborn_genetic_diseases|Parkinson_disease_8,_a...</td>\n",
       "      <td>rs34637584</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     avsnp150 Chr      Start        End Ref Alt Gene.refGene  AAchange  \\\n",
       "0  rs76763715   1  155205634  155205634   T   C          GBA   p.N409S   \n",
       "1  rs34637584  12   40734202   40734202   G   A        LRRK2  p.G2019S   \n",
       "\n",
       "                                         CLNSIG_1705  \\\n",
       "0  Pathogenic|other|other|Pathogenic|Pathogenic|P...   \n",
       "1                              Pathogenic|Pathogenic   \n",
       "\n",
       "                                 CLNSIG_1806  \\\n",
       "0  Pathogenic/Likely_pathogenic,_risk_factor   \n",
       "1                                 Pathogenic   \n",
       "\n",
       "                                         CLNDBN_1705  \\\n",
       "0  Gaucher's_disease,_type_1|Parkinson_disease,_l...   \n",
       "1  Parkinson_disease_8,_autosomal_dominant|not_pr...   \n",
       "\n",
       "                                         CLNDBN_1806         SNP  \n",
       "0  Dementia,_Lewy_body,_susceptibility_to|Rigidit...  rs76763715  \n",
       "1  Inborn_genetic_diseases|Parkinson_disease_8,_a...  rs34637584  "
      ]
     },
     "execution_count": 92,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# The common ones between potential pathogenics and Meta5\n",
    "df_pth = pd.read_csv('clinvar_1806_1705.csv')\n",
    "df_m5 = pd.read_csv('../../GWAS_prog/data/Meta5new.csv')\n",
    "df_m5 = df_m5[df_m5['Failed final filtering and QC']==0] # 90 SNPs\n",
    "pd.merge(df_pth, df_m5[['SNP']], left_on='avsnp150', right_on='SNP')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
